ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/KnownPeers.hs

```diff
@@ -356,13 +360,24 @@ aboveTarget PeerSelectionPolicy {
   , let numRootPeersCanForget = LocalRootPeers.size localRootPeers
                               + Set.size publicRootPeers
                               - targetNumberOfRootPeers
+        unfinishedConnections =
+          Map.keysSet
+          $ Map.filter (\case
+                          UnknownConnectionSt -> False
+                          _                   -> True
+                       )
+                       cmState
         availableToForget     = KnownPeers.toSet knownPeers
                                   Set.\\ EstablishedPeers.toSet establishedPeers
                                   Set.\\ LocalRootPeers.keysSet localRootPeers
                                   Set.\\ (if numRootPeersCanForget <= 0
                                             then publicRootPeers else Set.empty)
                                   Set.\\ inProgressPromoteCold
                                   Set.\\ bigLedgerPeers
+                                  -- We don't want to forget peers from an
+                                  -- async demotion that are still being
+                                  -- cleaned up.
+                                  Set.\\ unfinishedConnections
```

Comment:
I don't think this is the right approach. We're leaking the state from the `connection manager` to the `outbound-governor` and we duplicate the way `outbound governor` access it.  I think we should extend `PeerStatus` with `PeerCooldown` indicates that the peer is being demoted. Would it suffice for this use case?

Reply:
I get what you mean. The problem with what you suggested is that `PeerCold` doesn't mean the connection has effectively terminated. Tainting `PeerStatus` with an extra constructor seems wrong, because `PeerStatus` has a different purpose, I think it would make things complex. The only way to know if the connection has been cleaned is to share some state. It can be through `PeerStateActions` but you need the `peerconn` which will stop existing once the peer gets removed from the known peer set, which is the case in the test failure